### DISCUSSION TITLE
Mimic Windows 11 Flyouts with Acrylic Blur

### DISCUSSION THREAD

**Discussion by Cheerpipe** [AvaloniaUI/Avalonia#6857] [Q&A]
Hello. 

I am trying to mimic windows 11 flyout window and I've been doing well so far but i am struggling adding Acrylic blur.:

Acrylic Blur: `ShowInTaskbar="false"
                                CanResize="False"
                                Topmost="True"
                                SystemDecorations="None"
                                TransparencyLevelHint="AcrylicBlur"
                                ExtendClientAreaToDecorationsHint="True"
                                ExtendClientAreaChromeHints="NoChrome"
                                ExtendClientAreaTitleBarHeightHint="-1"
                                Background="Transparent"`
With this setting i have Acrylic Blur but i also have a square background that overflows my window content.

![image](https://user-images.githubusercontent.com/972765/140197605-19904807-5443-46bd-9714-41a312d9e9e4.png)

Also tried with system decoration Full: ` ShowInTaskbar="false"
                                CanResize="False"
                                Topmost="True"
                                SystemDecorations="Full"
                                TransparencyLevelHint="AcrylicBlur"
                                ExtendClientAreaToDecorationsHint="True"
                                ExtendClientAreaChromeHints="NoChrome"
                                ExtendClientAreaTitleBarHeightHint="-1"
                                Background="Transparent"`
![image](https://user-images.githubusercontent.com/972765/140198108-093fe74a-a694-491d-9e4c-f4b4f04d40a6.png)

Current implementation: `ShowInTaskbar="false"
                                CanResize="False"
                                Topmost="True"
                                SystemDecorations="None"
                                TransparencyLevelHint="Transparent"
                                ExtendClientAreaToDecorationsHint="True"
                                ExtendClientAreaChromeHints="NoChrome"
                                ExtendClientAreaTitleBarHeightHint="-1"
                                Background="Transparent"`

![b3CP6N5b7T](https://user-images.githubusercontent.com/972765/140197308-ecff2232-a669-4c9f-b646-e2816308d746.gif)

So my ultimate question is:

¿Is there any way to get an Acrylic Blur background clipped to a specific content or content?

This is what i trying to achieve btw:

![image](https://user-images.githubusercontent.com/972765/140198352-256f6005-0e17-4da9-b9a8-a022288ebef6.png)

**Comment by maxkatz6** [comment]
I don't know why there is a padding on the right, but I can guess about corners.
You are setting first window composition layer as a acrylic blur on top of which all your content is rendered. And most likely you also set CornerRadius on content itself. But there is nothing to tell that whole window should be with corner radius. And we don't have Window API for that either.
On other hand new Win32PlatformOptions.CompositionBackdropCornerRadius should be exactly what you need to round the corners. Though I am not sure right now if it's available in 0.10.10 or will be in 0.11.0.

**Reply by Cheerpipe** [reply]
First all, Thanks for your incredible fast and useful response!!!

> On other hand new Win32PlatformOptions.CompositionBackdropCornerRadius should be exactly what you need to round the corners.
Thanks. It is already implemented and working good so far.

![image](https://user-images.githubusercontent.com/972765/140240351-62789b61-a10a-40ff-a2c6-3389014372f3.png)


> I don't know why there is a padding on the right

It happens because the window is wider than the flyout content. The padding is what i am using to create "space" between the flyout and the display edge. I did it in this way because i am not animating the windows itself, i am just applyng a TranslateRenderTransform to simulate that the flylut start at the edge of the screen  to avoid this (The flyout is visible outside main screen if you have a dual screen setup):


![DEbnn4rtr8](https://user-images.githubusercontent.com/972765/140240574-57bb1ffc-8b9c-4208-8c87-ffda71899053.gif)

Before acrylic RenderTransform worked ok but it wont work with Acrylic because transform is not applied to the background:

![lnaYQXXUls](https://user-images.githubusercontent.com/972765/140240817-923b479c-45e7-4fcb-a6fe-ce5606138ceb.gif)


As a workaround i am moving the entire windows now and increasing the Width to simulate a windows clipping effect to keep it inside the main screen edge but the effect is not perfect as you can see:

![E474wMUlec](https://user-images.githubusercontent.com/972765/140241171-ce432e57-a20b-46c5-ab12-4aa4725ca7c0.gif)

It flashes on the second monitor (it happens because a window cant have les than 32 pixels width but i can manage this) and the drawing is not perfect, at least not with a 250ms animation.

Animation looks better with less speed (2000ms)

![kVP8ISNev3](https://user-images.githubusercontent.com/972765/140241355-17dbd18e-e37f-442d-bab0-f355e6449b6c.gif)

Is this a Windows API or Avalonia limiation? also, ¿Is ther a better way to clip a window with acrylic background? (also tried with ClipToBounds but it is the same. content is clipped but the acrylic background not).

Thanks in advance (and sorry for my english)

**Comment by Cheerpipe** [comment]
https://user-images.githubusercontent.com/972765/140252628-4e525ae0-fe7d-4e38-97ec-ab8f8cc68a03.mp4

This is best i can achieve moving the window and increasing the size of the windows while animating. The rendering fails in the right edge. It is like the background and window visuals are out of sync.

**Comment by Xerxes004** [comment]
I apologize for bumping this thread since it's pretty old. I just wanted to add some code to describe how to use the `Win32PlatformOptions` since I wasn't able to find documentation for it.

When setting up the app, the `AppBuilder` needed to have `UseWin32()` called like so in my `Program.cs`:

```C#
public static AppBuilder BuildAvaloniaApp()
    => AppBuilder.Configure<App>()
                 .UseWin32()
                 .UseReactiveUI()
                 .UsePlatformDetect()
                 .LogToTrace();
```

Right before I call `BuildAvaloniaApp()`, I added a lazy instantiation of `Win32PlatformOptions` to the `AvaloniaLocator` instance like so:

```C#
// Called in Main() right before creating the AppBuilder
// Fix for square corners
AvaloniaLocator.CurrentMutable.Bind<Win32PlatformOptions>()
               .ToLazy(() => new Win32PlatformOptions()
               {
                   CompositionBackdropCornerRadius = 7
               });
```

This removed the square corners for the window of my app in Windows 10.

I'm using Avalonia 11.0.0-preview4 at time of writing.

Cheers! 🍺

**Reply by Butcher-vj** [reply]
Hello @Xerxes004 , can you help me with AcrylicBlur effect in v11.0.11
I have used TransparencyLevelHint="AcrylicBlur"
Background="Transparent", which is working fine. But when i try to record my screen the transparency is flickering. In windows 11.
I searched in the web and they say like  .With(new Win32PlatformOptions
 {
     UseWindowsUIComposition = true
 })
 But it is not working on latest avalonia version.

**Comment by Zrylx** [comment]
I think the simplest way to add corner radius to a window directly is like so in AppBuilder (in context to the previous reply):
```
public static AppBuilder BuildAvaloniaApp()
    => AppBuilder.Configure<App>()
        .UsePlatformDetect()
        .With(new Win32PlatformOptions
        {
            WinUICompositionBackdropCornerRadius = 10
        })
        .WithInterFont()
        .LogToTrace();

```
